src/binary/ContractionOptimizer.cpp

```diff
@@ -0,0 +1,435 @@
+                    });
+  std::stable_sort( m_loops->begin(), m_loops->end(),
+                    [&](loop_property l_a, loop_property l_b) -> bool {
+                      return l_a.dim_type < l_b.dim_type;
+                    });
+  std::stable_sort( m_loops->begin(), m_loops->end(), 
+                    [&](loop_property l_a, loop_property l_b) -> bool {
+                      return l_a.exec_type < l_b.exec_type;
+                    });
+}
+
+void einsum_ir::binary::ContractionOptimizer::fuseLoops(){
+  for( std::vector<loop_property>::iterator l_it = m_loops->begin() + 1; l_it < m_loops->end(); l_it++ ){
+    std::vector<loop_property>::iterator l_other = l_it-1;
+    if( l_it->dim_type  == l_other->dim_type && 
+        l_it->exec_type == l_other->exec_type ){
+      int64_t l_size = l_it->size; 
+      if( l_it->stride_left    * l_size == l_other->stride_left    && 
+          l_it->stride_right   * l_size == l_other->stride_right   && 
+          l_it->stride_out     * l_size == l_other->stride_out     &&
+          l_it->stride_out_aux * l_size == l_other->stride_out_aux    ){
+        l_it->size *= l_other->size;
+        m_loops->erase(l_other);
+        l_it--;
+      }
+    }
+  }
+}
+
+void einsum_ir::binary::ContractionOptimizer::removeEmptyLoops(){
+  for( std::vector<loop_property>::iterator l_it = m_loops->begin(); l_it < m_loops->end(); l_it++ ){
+    if( l_it->size      == 1 &&
+        l_it->exec_type != exec_t::PRIM ){
+      m_loops->erase( l_it );
+    }
+  }
+}
+
+void einsum_ir::binary::ContractionOptimizer::moveLoopsToInternal(){
+  m_size_all_m = 1;
+  m_size_all_n = 1;
+  m_free_loops.resize(5);
+  std::vector<loop_property>::iterator l_it = m_loops->begin();
+  while( l_it < m_loops->end() ){
+    if( l_it->exec_type == exec_t::SEQ ){
+      if( l_it->dim_type == dim_t::M ){
+        m_size_all_m *= l_it->size;
+      }
+      if( l_it->dim_type == dim_t::N ){
+        m_size_all_n *= l_it->size;
+      }
+      m_free_loops[l_it->dim_type].push_back(*l_it);
+      m_loops->erase(l_it);
+    }
+    else{
+      l_it++;
+    }
+  }
+}
+
+void einsum_ir::binary::ContractionOptimizer::addKernel(){
+  typedef enum {
+    PRIM_BR = 0,
+    PRIM_C  = 1,
+    PRIM_M  = 2,
+    PRIM_N  = 3,
+    PRIM_K  = 4
+  } gemm_prim_t;
+
+  //find possible kernel dimensions
+  bool l_transpose_a = false;
+  bool l_transpose_b = false;
+  int64_t l_potential_kernel_size[] = {1,1,1,1,1};
+  std::vector<loop_property>::iterator l_potential_kernel_loop[] = { m_free_loops[ dim_t::K ].end(),
+                                                                     m_free_loops[ dim_t::C ].end(),
+                                                                     m_free_loops[ dim_t::M ].end(),
+                                                                     m_free_loops[ dim_t::N ].end(),
+                                                                     m_free_loops[ dim_t::K ].end() };  
```

Comment:
Does this introduce a dependency on the values in the ``dim_t`` enums?
In other words: If I change the IDs of the types, will the code still work?

Refactor if not.